src/Integration.Vsix/CFamily/Protocol.cs

```diff
@@ -0,0 +1,328 @@
+        public long MsVersion { get; set; }
+        public string[] IncludeDirs { get; set; } = Array.Empty<string>();
+        public string[] FrameworkDirs { get; set; } = Array.Empty<string>();
+        public string[] VfsOverlayFiles { get; set; } = Array.Empty<string>();
+        public string ModuleName { get; set; } = "";
+        public string Predefines { get; set; } = "";
+        public string[] Macros { get; set; } = Array.Empty<string>();
+        public string TargetTriple { get; set; } = "x86_64-unknown-unknown";
+        public string File { get; set; } = "";
+    }
+
+    internal class Response
+    {
+        public Message[] Messages { get; }
+
+        public Response(Message[] messages)
+        {
+            Messages = messages;
+        }
+    }
+
+    internal class MessagePart
+    {
+        public string Filename;
+        public int Line { get; }
+        public int Column { get; }
+        public int EndLine { get; }
+        public int EndColumn { get; }
+        public string Text { get; }
+
+        public MessagePart(string filename, int line, int column, int endLine, int endColumn, string text)
+        {
+            Filename = filename;
+            Line = line;
+            Column = column;
+            EndLine = endLine;
+            EndColumn = endColumn;
+            Text = text;
+        }
+    }
+
+    internal class Message : MessagePart
+    {
+
+        public string RuleKey { get; }
+        public int RemediationCost { get; }
+        public bool PartsMakeFlow { get; }
+        public MessagePart[] Parts { get; }
+
+        public Message(string ruleKey, string filename, int line, int column, int endLine, int endColumn, int remediationCost, string message, bool partsMakeFlow, MessagePart[] parts)
+            : base(filename, line, column, endLine, endColumn, message)
+        {
+            RuleKey = ruleKey;
+            RemediationCost = remediationCost;
+            PartsMakeFlow = partsMakeFlow;
+            Parts = parts;
+        }
+
+    }
+
+    internal static class Protocol
+    {
+        /**
+          * This method does not close the provided stream.
+          */
+        public static void Write(BinaryWriter writer, Request request)
+        {
+            WriteUTF(writer, "IN");
+            Write(writer, request.Options);
+            WriteLong(writer, request.Flags);
+            WriteLong(writer, request.MsVersion);
+            Write(writer, request.IncludeDirs);
+            Write(writer, request.FrameworkDirs);
+            Write(writer, request.VfsOverlayFiles);
+            WriteUTF(writer, request.ModuleName);
+            WriteUTF(writer, request.Predefines);
+            Write(writer, request.Macros);
+            WriteUTF(writer, request.TargetTriple);
+            WriteUTF(writer, request.File);
+            WriteUTF(writer, "END");
+        }
+
+        public /*visible for testing*/ static void WriteLong(BinaryWriter writer, long l)
```

Comment:
The convention in this repo is `internal /* for testing */`.

As a rule of thumb, it's the only time a `internal` is used on a method.
If a class is only to be used internally then the class is marked `internal`, and any methods that are expected to be called by other types in the assembly are marked `public`.